infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -275,13 +306,15 @@ private LogData readRecord(FileHandle fh, long address)
                     }
                 }
 
-                if (address == -1) {
+                if (isOverwritten(fh, filePosition)) {
+                    log.debug("Trim at {} hit, reading (size={}), position {}.", address, metadata.getLength(), filePosition);
+                } else if (address == -1) {
                     //Todo(Maithem) : maybe we can move this to getChannelForAddress
-                    fh.knownAddresses.add(entry.getGlobalAddress());
+                    fh.knownAddresses.put(entry.getGlobalAddress(), filePosition);
```

Comment:
@Maithem the only thing I would worry about here is atomicity, since the steamLog and the overwriteAddresses log are updates separately.

Do you see any potential issue here?